### ISSUE TITLE
Add builds for MariaDB

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-framework#8655]
MariaDB was forked off MySQL around MySQL 5.5, and ported features from MySQL 5.6. There are some [incompatibilities](https://mariadb.com/kb/en/library/incompatibilities-and-feature-differences-between-mariadb-103-and-mysql-57/) between MariaDB 10.3 (latest stable) and MySQL 5.7. I don't think we would be affected by our fairly basic usage of the databases, but given the relative popularity of MariaDB we should start testing on it and guard against those. MariaDB is [available](https://docs.travis-ci.com/user/database-setup/#mariadb) via Travis.

I would suggest that we fold the MariaDB build into another build variation, to avoid bloating our build matrix further. Alternatively, we could mark it as a nightly build only.

Context: Discussion around [supporting MySQL 5.5](https://github.com/silverstripe/silverstripe-framework/issues/8651)

/cc @silverstripe/core-team @dnsl48

**Comment by maxime-rainville** [comment]
I'm in favour of anything that gives me the sentiment of sticking it to Larry Ellison and Oracle.

Would be interesting to see if they are any performance difference between the two.

**Comment by dhensby** [comment]
It would be good to add, possibly as an allowed_failure so that builds can finish earlier and also because we don't provide explicit support for maria db

**Comment by chillu** [comment]
Let's see where we stand with builds. I think given the relatively small differences between MySQL and MariaDB in our context, the overhead of claiming MariaDB support is minimal.

**Comment by dnsl48** [comment]
What about the versions? Should we go after both `5.5` and `10`?
https://mariadb.org/about/maintenance-policy/

**Comment by sminnee** [comment]
I would suggest that we test MariaDB 10, and skip 5.5

**Comment by dhensby** [comment]
> If we start testing it, then I think we should claim its support.
> If we claim the support, then we should'n use `allowed_failure` I reckon.

That seems like a poor use of logic. None of those things actually mandate the other. We test older PHP versions because we can, not because we support them. Allowing failures so that we can get quicker success statuses for builds makes sense to me.

> I think given the relatively small differences between MySQL and MariaDB in our context, the overhead of claiming MariaDB support is minimal.

If the overhead is minimal (but still non-zero) then what's the harm in *not* claiming support?

**Comment by dnsl48** [comment]
> We test older PHP versions because we can, not because we support them
In another thread people told that we actually support the old versions. Sorry, but to me that looks like a waste of CI resources if we run tests which we don't care about. If we don't want something to be broken, then we shouldn't allow tests to break, otherwise what's the point of running those tests?
On the other hand, fixing something that's been broken for a long time is usually much harder than something freshly broken. That's the whole point of CI in my opinion.

**Comment by sminnee** [comment]
I tend to agree with @dnsl48 on misgivings around allowed_failures. When we’ve had them in the past - HHVM or PHP nightlies? - they’ve mainly been left to rot with little impact on the behaviour of PR submitters / mergers. I’ve tended to take the approach “either it breaks the build or it’s not worth running”

I do wonder if we need a “2-speed build”, though, with a nightly build for example? Adding different versions of databases and mariaDb to test - which I think are important to treat as breaks - will bloat out our builds and I’m not sure we need to run that on every PR

**Comment by dhensby** [comment]
How does a nightly build change the behaviour of submitters or mergers in a way that allowed failures don't?

What allowed_failures do is say "this is not ideal, but it's non-critical", like a warning in a linter.

Anyway, I was suggesting a way to get visibility over the state of something unsupported without slowing down build times (because allowed failures combined with early reporting of status means it won't impact individual build times). Obviously it's a bad idea if we are saying we are 100% supporting maria DB as a first class dependency

**Comment by sminnee** [comment]
Yeah I don’t think we should be discouraging people from using MariaDB over MySQL by treating it as less supported. So I would advocated adding a recent version of MariaDB to the supported list.

**Comment by sminnee** [comment]
Yeah, daily builds or allowed_failures are equally useful in providing extended builds if and only if you have someone paying attention to them outside the PR merge cycle. Daily builds are a bit cheaper but not fundamentally different, I admit.

**Comment by sminnee** [comment]
I would suggest running builds against 10.0 and 10.3 and reason that 10.1 and the 10.2 probably won’t break if those two work.

For MYSQL I think we should run against 5.6, 5.7, 8.0.

We don’t need any more than 1 build against each of these so hopefully we can add this in without too much buildcount inflation

**Comment by sminnee** [comment]
It seems like it's quite difficult to run MySQL 8 on Travis, which is a bit surprising, so I wonder if we focus our testing on MyQL 5.6 and 5.7... How many people are running MySQL 8 I wonder?

**Comment by sminnee** [comment]
Since this is about adding builds; this was another PR that was manipulating the build matrix #8503. I added SQLite builds, and ensured that MySQL/SQLite/PGSQL were all run in both PDO and non.

**Comment by dnsl48** [comment]
I think we can run `MySQL 8` in a Docker container.
Travis supports docker nicely nowadays: https://docs.travis-ci.com/user/docker/

**Comment by chillu** [comment]
I would suggest that we add MariaDB 10 to the framework build, since that'll provide the most coverage in terms of ORM. I'm hesitant adding it to *all* of our 80ish supported modules though, because of the overhead of actually sending all those `travis.yml` pull requests, but also because it'll increase overall build time. 

The "update dozens of travis.yml files" problem will happen all the time though, e.g. when we start testing with PHP 7.3. Maybe we just need to bite the bullet and script that? "Foreach supported module, do a grep in the travis.yml, send a PR, list the PR URL". Then you'd need to manually check if the builds have passed on all of those PRs, and press the shiny green button on a whole bunch of browser tabs.
